### ISSUE TITLE
MSVC minor and build versions

### ISSUE THREAD

**Issue by Lnk2past** [conan-io/conan#3311]
I searched around but I could not find anything, so please forgive me if I missed something.

Currently Conan supports the ability to specify the compiler version and compiler toolset to use for building. This works fine as is, but unfortunately has become insufficient with how Microsoft treats minor and patch releases for Visual Studio.

Knowing that a particular library was built with VC15 is no longer enough, and even further it is still insufficient knowing it was built with VC15.7. It is necessary to know what specific build of MSC was used to build a given package.

Here is a recent example illustrating the issue:

- Visual Studio 15.7.0 is not binary compatible [in some areas] with 15.6.x when using C++17
- Visual Studio 15.7.3 resolves this (including the patch from 15.7.1)
- Visual Studio 15.7.5 resolves a critical bug with NaN operations when using optimization
- Visual Studio 15.7.6 introduces no changes to the compiler

Not even consider 15.7.1, 15.7.2, and 15.7.4 here, we have a number of _different compilers_ used for building code that are producing _different binaries_. There are a lot of issues here. When I get a package specifying VC15, am I getting something that is bugged? Am I getting something that actually isn't binary compatible with my project? _Am I getting something I can actually use?_

More so, when moving between versions that are irrelevant, it is not entirely useful to say that something was built with 15.7.5 vs 15.7.6, because they are using the same compiler build. The actual MSC full version build number should be used here:

15.7.5 and 15.7.6 use 191426433
15.7.3 and 15.7.4 use 191426430
etc.

This information can be scraped from the compiler itself; these are the MSC FULL VER values within the compilers.

To summarize:

- Minor and patch versions in Visual Studio matter
- Conan needs to be aware of them
- You cannot have multiple patch versions installed side by side on the same system (at least not without a sledge hammer), and as such Conan should not allow users to specify the minor/patch version of the compiler within the settings (i.e. settings.yml _should not change_)
- Conan should warn users when using a package built with a different patch version
- Conan should track the compiler patch version using the MSC FULL VER values.

I may have missed something here, so if anyone wants to chime in and provide any corrections, please do so. If I missed something within Conan/the documentation, please let me know.

**Comment by Lnk2past** [comment]
Hi @memsharded, thanks for the response. To be clear, I am not proposing that Conan do much else other than embed the actual compiler version into a package's metadata. The problem(s) ultimately lay with Microsoft, not Conan. However, Conan should be able to provide enough information about packages to allow users to act smarter as needed.

As I noted in my OP, I do not think settings.yml should change here. Users absolutely **should not** be doing anything like ```conan create ... -s compiler.version="191426433"```. Given that you cannot really have multiple versions of VC15 installed side by side anyway, I wouldn't expect Conan to support this. As such, I do not think package ID's should take this into account.

I agree that package providers should not be creating a package for each minor version/patch of the compiler. All I ultimately would like is for Conan to add an additional piece of information detailing the exact compiler version used to build the package. This way, I know and other teams know what to expect when we get some package from our central store, whether it is a 3rd party library or an internally developed one.

**Comment by Lnk2past** [comment]
@memsharded, any more thoughts on this? All that I am asking for is for Conan to include the exact compiler build version with the package (**not** in-place of whatever is in ```compiler.version```). That way we could search for/filter for packages built with particular compiler builds within our central store, accessible as a part of its metadata.

**Comment by Lnk2past** [comment]
Conan would capture that information automatically whenever a package is built using Visual Studio; this is not something that would go in ```settings.yml``` and should not be found anywhere in the recipes. Users shouldn't have control of this additional information as it is purely informational. They can ignore it if they like, but for those that may require having that level of insight into their packages should be able to readily see this information.

Settings, options, and requirements are already made available in the metadata of a package (as can be seen [here when using Artifactory](https://www.jfrog.com/confluence/display/RTF/Conan+Repositories)). Couldn't there be another section in the metadata that could include this information?

If I can boil down my request a little more, this is what we need:

- When building with VS, Conan should record the exact compiler build version when building with VS
- Conan should provide this information as an additional piece of metadata within the package

**Comment by memsharded** [comment]
Ok, I think I understand the issue. However, it is something that looks extremely complex: that means adding something new to the model, which affect both the client and servers, and a whole new flow of information: build helpers need to capture it somewhere (some text file within the packages). Those files will be formatted and can be parsed by servers to display that information. The format of those files would be quite difficult to  find something that makes sense to most users. Which information, to which level should be logged there? Because if storing the exact compiler number for VS sounds reasonable, then why not all the metadata of the compilers too? Maybe some information about the machines that did the built? 

So even if the feature request could make sense, it is indeed something that cannot be approached without thinking about the general use case of capturing metadata about the builds, and its later usage. This is a big investment, and cannot be put in the short term roadmap. I will leave it open for discussion, seeing if it gets support from other users, and further feedback from the community.

**Comment by jgsogo** [comment]
Hi, @Lnk2past , in this issue https://github.com/conan-io/conan/issues/3573 there is a lot of valuable feedback about using the `toolset` instead of the IDE version, and we are really thinking about adding that as a new compiler to `Conan v1.x`. I think you will find that issue very interesting.

**Comment by Lnk2past** [comment]
While I am not longer really using Conan for anything any more (new job and all, hardly any more C++ for me 😢), I think as long as there is a finer grained distinction between MSVC compiler versions then the issue mentioned here would be taken care of. If the other issue has more traction and an agreed upon path forward, then go ahead and close this (unless you think it still holds value to keep open). Thanks for the follow up!

**Comment by anton-danielsson** [comment]
I stumbled upon another case where this breaks:
There's a hard ABI break in 16.5 caused by:
https://docs.microsoft.com/en-us/cpp/overview/cpp-conformance-improvements?view=msvc-160#tls-guard-changes
It will cause errors like:
`error LNK2019: unresolved external symbol __dyn_tls_on_demand_init`
Saw this with protobuf but it affects all code that use `thread_local` variables.
So it's not enough to have just `16` as compiler version since a library built with
16.5 can't be used by an application built with 16.4 (or less).

**Comment by memsharded** [comment]
Hi @anton-dirac 

Have you checked the new ``msvc`` compiler setting? https://github.com/conan-io/conan/blob/c5abf1a463e121f84b981bd9029122114227ec99/conans/client/conf/__init__.py#L87

This is intended to replace the ``Visual Studio`` setting in Conan 2.0. You can see that this contain the compiler version down to the minor, so it should probably be ok. Conan default detection could be the "major" one like ``"19.1"`` which would work for most users, and users wanting to have finer control could use the full one, like ``"19.11"`` or ``"19.12"``.

The suggestion would be:
- If Conan is using the "major" setting msvc ``"19.1"``, then it is necessary that users generally upgrade to the minor VS version used to create that package.
- If different packages are desired for different minors ``"19.11"``, then different users can work with different VS minor versions , as long as they define it in their profiles (and someone is building the package binaries for that minor VS version)

**Comment by anton-danielsson** [comment]
Wow, this looks very promising!

Did some quick tests and it seems that all the checks are not in place yet?
I had `compiler.version=19.28` in my profile but I used 19.23. (`The C compiler identification is MSVC 19.23.28106.4`)
I guess the `check_compiler_version` stuff is still work-in-progress?

**Comment by memsharded** [comment]
Yes, correct. The ``check_compiler_version`` will likely be implemented into the ``CMakeToolchain()``. The ``msvc`` compiler is only managed by the new ``conan.tools.xxxx`` helpers, not the previous ones, and it is not intended to do it, it will only move forward towards Conan 2.0 into the new tools.

**Comment by Lnk2past** [comment]
With the closing of #3573 I think this is OK to close. I have not been involved/using Conan in a long while now, and so I do not have anything else to add here. Glad to see this feature move forward though!
